core/src/main/java/io/grpc/internal/AbstractClientStream2.java

```diff
@@ -0,0 +1,300 @@
+  }
+
+  @Override
+  public final void halfClose() {
+    if (!outboundClosed) {
+      outboundClosed = true;
+      endOfMessages();
+    }
+  }
+
+  @Override
+  public final void cancel(Status reason) {
+    Preconditions.checkArgument(!reason.isOk(), "Should not cancel with OK status");
+    cancelled = true;
+    abstractClientStreamSink().cancel(reason);
+  }
+
+  @Override
+  public final boolean isReady() {
+    return super.isReady() && !cancelled;
+  }
+
+  /** This should only called from the transport thread. */
+  protected abstract static class TransportState extends AbstractStream2.TransportState {
+    /** Whether listener.closed() has been called. */
+    private boolean listenerClosed;
+    private ClientStreamListener listener;
+
+    private Runnable deliveryStalledTask;
+
+    private boolean headersReceived;
+    /**
+     * Whether the stream is closed from the transport's perspective. This can differ from {@link
+     * #listenerClosed} because there may still be messages buffered to deliver to the application.
+     */
+    private boolean statusReported;
+
+    protected TransportState(int maxMessageSize) {
+      super(maxMessageSize);
+    }
+
+    @VisibleForTesting
+    public final void setListener(ClientStreamListener listener) {
+      Preconditions.checkState(this.listener == null, "Already called setListener");
+      this.listener = Preconditions.checkNotNull(listener, "listener");
+    }
+
+    @Override
+    public final void deliveryStalled() {
+      if (deliveryStalledTask != null) {
+        deliveryStalledTask.run();
+        deliveryStalledTask = null;
+      }
+    }
+
+    @Override
+    public final void endOfStream() {
+      deliveryStalled();
+    }
+
+    @Override
+    protected final ClientStreamListener listener() {
+      return listener;
+    }
+
+    /**
+     * Called by transport implementations when they receive headers.
+     *
+     * @param headers the parsed headers
+     */
+    protected void inboundHeadersReceived(Metadata headers) {
+      Preconditions.checkState(!statusReported, "Received headers on closed stream");
+      headersReceived = true;
+      listener().headersRead(headers);
+    }
+
+    /**
+     * Processes the contents of a received data frame from the server.
+     *
+     * @param frame the received data frame. Its ownership is transferred to this method.
+     */
+    protected void inboundDataReceived(ReadableBuffer frame) {
+      Preconditions.checkNotNull(frame, "frame");
+      boolean needToCloseFrame = true;
+      try {
+        if (statusReported) {
+          log.log(Level.INFO, "Received data on closed stream");
+          return;
+        }
+        if (!headersReceived) {
+          transportReportStatus(
+              Status.INTERNAL.withDescription("headers not received before payload"),
+              false, new Metadata());
+          return;
+        }
+
+        needToCloseFrame = false;
+        deframe(frame, false);
```

Comment:
Remind me... If deframing fails for some reason, does the deframer close the frame?